Stats Improvements Step 3: add scrubbing interaction

anitaa1990 opened the pull request:
Fixes #1252 . This PR completes the second step of the Stats Improvements Project as defined in the master thread [here](https://github.com/woocommerce/woocommerce-android/issues/1199).

### Changes
- [x] Update the fluxC hash that includes the visitor stats changes needed for scrubbing interaction.
- [x] Make modifications to the old stats UI implementation to fetch visitor stats data from local cache.
- [x] Make modifications to the new v4 stats UI implementation to fetch visitor stats data from local cache.
- [x] Add scrubbing interaction to v4 stats UI. Since the visitor stats api does not support hourly data breakup, during scrubbing of `Today` tab, hide the view instead.
- [x] Add scrubbing interaction to old stats UI.
- [x] Old Stats UI: Update the date picker when scrubbing takes place. The date would be formatted based on the granularity. 
- [x] New Stats UI: Update the date picker when scrubbing takes place. The date would be formatted based on the granularity. 
- [x] Fixing a bug where the visitor stats for `This Year` is not updated during scrubbing
- [x] Fixing a bug where the visitor stats was not displayed when pull to refresh was called
- [x] I also added an extension class for formatting dates used in the scrubbing. This was because the `DateUtils` class was being too messy with a lot of code and I was getting confused 😕 

### Behaviour
(LEFT: Old Stats UI . RIGHT: New Stats UI)

#### DAYS/Today Tab:
<img width="300" src="https://user-images.githubusercontent.com/22608780/62757613-7bb91300-ba99-11e9-8393-a7e0a45b2ef8.gif"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/62757614-7d82d680-ba99-11e9-96d8-885a93c56924.gif">

#### WEEKS/This Week Tab:
<img width="300" src="https://user-images.githubusercontent.com/22608780/62757724-e8341200-ba99-11e9-8450-238d737ea570.gif"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/62757707-d8b4c900-ba99-11e9-9443-65bab26cbd71.gif">

#### MONTHS/This Month Tab:
<img width="300" src="https://user-images.githubusercontent.com/22608780/62757855-3ba66000-ba9a-11e9-84d3-684e4235733e.gif"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/62757863-3d702380-ba9a-11e9-8d92-72fd4eb672a8.gif">

#### YEARS/This Year Tab:
<img width="300" src="https://user-images.githubusercontent.com/22608780/62757943-89bb6380-ba9a-11e9-8e31-3ff97fdc5cea.gif"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/62757952-90e27180-ba9a-11e9-99c1-8e93bdde9779.gif">

### Screenshots
(LEFT: Old Stats UI . RIGHT: New Stats UI)

#### DAYS/Today Tab:
<img width="300" src="https://user-images.githubusercontent.com/22608780/62848806-b7064c80-bcfa-11e9-9fcd-7cf3f3e81f82.png"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/62848811-b968a680-bcfa-11e9-8cd7-d55d9dcce288.png">

#### WEEKS/This Week Tab:
<img width="300" src="https://user-images.githubusercontent.com/22608780/62848861-ec129f00-bcfa-11e9-8bdc-555b53e2a98d.png"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/62848863-ee74f900-bcfa-11e9-9bd2-c5be78000535.png">

#### MONTHS/This Month Tab:
<img width="300" src="https://user-images.githubusercontent.com/22608780/62848881-064c7d00-bcfb-11e9-9bc0-75ae7e947c99.png"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/62848884-09476d80-bcfb-11e9-83f4-483a85a65053.png">

#### YEARS/This Year Tab:
<img width="300" src="https://user-images.githubusercontent.com/22608780/62848900-17958980-bcfb-11e9-9687-2572adb48c36.png"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/62848903-1a907a00-bcfb-11e9-9e58-d7e3db097405.png">

anitaa1990 committed:
Update FluxC hash to support visitor stats scrubbing interaction

anitaa1990 committed:
Current Stats UI - Update the local visitor stats code

anitaa1990 committed:
V4 Stats UI - Update the local visitor stats code

anitaa1990 committed:
V4 Stats UI - Added scrubbing interaction

anitaa1990 committed:
Added separate extension class to format date string since the DateUtils class was getting too cluttered

anitaa1990 committed:
Old Stats UI - Added scrubbing interaction

nbradbury commented:
@anitaa1990 I realize this is a draft, but thought I'd point out that the scrubbing isn't very reliable. Sometimes I can swipe between bars and it works fine, other times I swipe and nothing happens.

Also, if I lift my finger after swiping and then tap in an empty area, I would expect the chart to deselect the selected bar and show the stats for the chosen date range. Instead, tapping an empty area selects the bar below it. So there's no obvious way to return to seeing the entire stats for the chosen date range.

anitaa1990 committed:
disable parent scrollview when the chart data is not null

anitaa1990 committed:
Fix failing test in DateUtils

anitaa1990 commented:
Thank you very much for testing @nbradbury!!

> Sometimes I can swipe between bars and it works fine, other times I swipe and nothing happens.

Ah, this was annoying me for a while now and was because the chart is inside a scrollview and the scroll is intercepted by the parent. I fixed this in 42690a1

> if I lift my finger after swiping and then tap in an empty area, I would expect the chart to deselect the selected bar and show the stats for the chosen date range. Instead, tapping an empty area selects the bar below it. 

This one is a little bit more tricky. Reading some similar issues from [here](https://github.com/PhilJay/MPAndroidChart/issues/3627) and [here](https://stackoverflow.com/questions/38717676/how-to-identify-bar-click-event-in-mpandroidchart), it seems this is not directly supported. We could set the `setMaxHighlightDistance` for the chart but this is not customisable for each bar but for the whole graph. We receive the callback event for the click though in `onValueSelected`. I'm still exploring this and if we can find some way to know where the click happened, we could just handle it in the callback

nbradbury commented:
> this was annoying me for a while now and was because the chart is inside a scrollview and the scroll is intercepted by the parent. I fixed this in 42690a1.

**Much** better!

> This one is a little bit more tricky...it seems this is not directly supported.

Actually, this may not be a big deal after all. I noticed that if you tap a bar that's already highlighted, it deselects the bar and returns to showing all the stats for the period. That pretty much addresses my concern about not being able to easily deselect.

anitaa1990 committed:
Fix lint issues

anitaa1990 committed:
Fixed a bug where the visitor stats data was not updated during scrubbing

anitaa1990 committed:
V4 stats UI - Added methods to parse date based on granularity during scrubbing

anitaa1990 committed:
V4 stats UI - Added logic to update date bar during scrubbing

anitaa1990 committed:
Old Stats UI - update date bar when scrubbing takes place

anitaa1990 committed:
New Stats UI - show the visitor stats layout when pull to refresh is called

anitaa1990 commented:
I feel like the month in the date bar differs between long and short (August vs Aug). I'm not sure if that is a bit inconsistent 🤔 

@nbradbury @kyleaparker could I trouble you for your review on ☝️ please? Thank you!!!

nbradbury commented:
@anitaa1990 I'm really loving these changes! 

I think it would be nice to be consistent with the dates both in the date bar and the marker view (the little box with the arrow that appears on a bar when you tap on it). I'm leaning towards short month names, but don't have a strong opinion about it.

I think we also need to be consistent with formatting the day value. In the screenshot below the date bar has "07" for the day while the marker has "7" - I definitely prefer the latter.

Also, when you have a chance can you pull `develop` into this? When I switch to this branch I have to re-install the app because it's older than `develop`.

![Screenshot_1565345047](https://user-images.githubusercontent.com/3903757/62771725-dfc4e280-ba6b-11e9-9ac3-45783c380854.png)

anitaa1990 committed:
Merge branch 'feature/stats-improvements-ui-changes' into anitaa/issue-1252-v4-scrubbing-changes

anitaa1990 commented:
@nbradbury Thank you!! 

> can you pull develop into this? When I switch to this branch I have to re-install the app because it's older than develop.

Sorry! I've been meaning to do that and it completely slipped my mind. I've merged the changes now. 

> I think it would be nice to be consistent with the dates both in the date bar and the marker view (the little box with the arrow that appears on a bar when you tap on it). I'm leaning towards short month names, but don't have a strong opinion about it.

I liked the long version because it looked good in the designs when displayed at the date bar but I noticed we use the shorter version throughout the old stats, in the x-axis and the marker view. I'm good either way. @kyleaparker WDYT? 

I think we also need to be consistent with formatting the day value. In the screenshot below the date bar has "07" for the day while the marker has "7" - I definitely prefer the latter.

Ah, good point! I didn't notice this. I'll change this to 7 instead of 07 later today 👍

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/extensions/DateExt.kt:
```diff
@@ -0,0 +1,22 @@
+package com.woocommerce.android.extensions
```
I'm getting four inspection warnings here:

<img width="1026" alt="Screen Shot 2019-08-09 at 11 51 01 AM" src="https://user-images.githubusercontent.com/3903757/62791929-2cc0ad00-ba9c-11e9-95dd-5aa5b2e49296.png">

anitaa1990 replied:
👍 Done in 4cb8a07

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/extensions/DateFormatExt.kt:
```diff
@@ -0,0 +1,78 @@
+package com.woocommerce.android.extensions
```
This might be better named `StringDateFormatExt` since all the functions extend `String` rather than `Date`.

anitaa1990 replied:
👍 Done in 38bfe77

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/extensions/DateFormatExt.kt:
```diff
@@ -0,0 +1,78 @@
+ * Method to convert date string from yyyy-MM-dd format to yyyy-MM format
+ */
+fun String.formatDateToYearMonth(): String {
+    val (year, month, day) = this.split("-")
```
These three functions trust the string to be in the desired date format, which I'm not sure is safe to do. I believe we've had cases where an endpoint gave us a date in an unexpected format.

anitaa1990 replied:
Good point! Fixed in 38bfe77

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/dashboard/DashboardStatsBarChart.kt:
```diff
@@ -46,4 +47,10 @@ class DashboardStatsBarChart(context: Context?, attrs: AttributeSet?) : BarChart
         }
     }
+
+    override fun onTouchEvent(event: MotionEvent): Boolean {
```
np: It would be nice to have a comment here explaining why this is necessary.

anitaa1990 replied:
👍 Done in 92f11af

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/mystore/MyStoreStatsView.kt:
```diff
@@ -295,38 +376,32 @@ class MyStoreStatsView @JvmOverloads constructor(ctx: Context, attrs: AttributeS
-    fun showVisitorStatsError() {
-        fadeInLabelValue(visitors_value, "?")
-    }
+    private fun getFormattedVisitorValue(date: String) =
```
I _think_ our convention for functions with inferred return types is that they should be one or two lines.

anitaa1990 replied:
👌 Done in e059f04

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/mystore/MyStoreStatsView.kt:
```diff
@@ -295,38 +376,32 @@ class MyStoreStatsView @JvmOverloads constructor(ctx: Context, attrs: AttributeS
     private fun fadeInLabelValue(view: TextView, value: String) {
         // do nothing if value hasn't changed
-        if (view.text.toString().equals(value)) {
+        if (view.text.toString() == value) {
```
👍

nbradbury commented:
I left some nitpicky-ish comments, but all in all this looks great @anitaa1990! I think we'll be ready to :shipit: Monday.

Also, it may be best to ask for design review when the main feature branch is ready. That way @kyleaparker can do a single, broader design review where she can see everything all at once (@kyleaparker correct me if I'm wrong!)

anitaa1990 committed:
Added exception handling to date format methods to handle scenarios with incorrect date formats

anitaa1990 committed:
Added comment to onTouchEvent method in custom BarChart class

anitaa1990 committed:
Modified inferred return types for methods with lines exceeding 2 lines

anitaa1990 committed:
Added explicit return type to methods in `DateExt` classes to address inspection warnings

anitaa1990 committed:
Modified methods to format date when scrubbing to match the marker view dates

anitaa1990 commented:
Thank you for the quick review @nbradbury! 

I've modified the dates both in the date bar and the marker view to display consistent data, fixed the nit picks in the code and updated the latest screenshots to the PR description. Ready for another round 👍

nbradbury commented:
Looks great! :shipit:

kyleaparker commented:
>I liked the long version because it looked good in the designs when displayed at the date bar but I noticed we use the shorter version throughout the old stats, in the x-axis and the marker view. I'm good either way. @kyleaparker WDYT?

Let's use the short version of the month on the today and week view and always include the 0 before single digit dates. We can use the full month when on the month view. 

We also no longer need the marker popover since the scrubber functionality is already showing the sales stats for the selected bar

nbradbury commented:
@anitaa1990 Looks like Kylea added [that comment](https://github.com/woocommerce/woocommerce-android/pull/1334#issuecomment-520457663) after I merged the PR. Feel free to make the change in a separate PR.
